src/validate.js

```diff
@@ -0,0 +1,36 @@
+import { isObject, toErrorList, mergeObjects } from "./utils";
+
+
+function createErrorHandler(formData) {
+  const handler = {
+    __errors: [],
+    addError(message) {
+      this.__errors = [...this.__errors, message];
```

Comment:
This, and the definition of `reduce` to return new objects every iteration, feel a little scary to me because they won't be as efficient as mutating objects in-place, and I'm not sure the JS RTS is smart enough to optimize the update into an in-place mutation. I know this is JS and we probably don't care about the overhead of copying objects, but it just feels wasteful to me and not very idiomatic. But JS isn't my best language so I'm willing to accept whatever here.